### PR TITLE
Pawn valid move

### PR DESCRIPTION
**Pull request by AlekstheRealist**
- Did a lot of shit, please take your time 

- Added boat load of tests and passed all of them!

- Added code. BAM.

### CODE REVIEW

**Comment by ruishine** [comment]
Hey guys! this is SO GREAT!! I'm so glad you guys worked on this and it passes all the tests!

The comments I'm about to make next are purely for refactoring purposes and I will do them by line.  Thanks for putting so much work into this!

File: app/models/pawn.rb [hunk 1]
```diff
@@ -1,5 +1,76 @@
...
+    elsif self.y_cord == 6 && self.color == 'black'
+      return true
+    else
+      return false
+    end
+  end
+
+  #Checks whether or not position pawn is moving to is backwards
+  def not_backwards?(x, y)
+    if self.color == 'white' && y > self.y_cord
```
Comment (ruishine) [inline app/models/pawn.rb]: I think this method can be taken out if we do the math without absolute values and just have new_y - y_cord == in the other methods

File: app/models/pawn.rb [hunk 2]
```diff
@@ -1,5 +1,76 @@
...
+  end
+
+  #Checks that pawn is on the originating square aka first_move_position depending on its color
+  def first_move_position?
+    if self.y_cord == 1 && self.color == 'white'
+      return true
+    elsif self.y_cord == 6 && self.color == 'black'
+      return true
+    else
+      return false
```
Comment (ruishine) [inline app/models/pawn.rb]: I'm not sure, but maybe try to take out all the returns.  like the if statement will return true automatically if the statement is true.  Not sure for else.  But there's at least an extra else, return false.

File: app/models/pawn.rb [hunk 3]
```diff
@@ -1,5 +1,76 @@
...
+  def two_squares_forward?(x, y)
+    first_move_position? && not_backwards?(x, y) && (self.y_cord - y).abs == 2 && self.x_cord - x == 0
+  end
+
+  def one_square_forward?(x, y)
+    not_backwards?(x, y) && (self.y_cord - y).abs == 1 && self.x_cord - x == 0
+  end
+
+  def diagonal_move?(x, y)
+    diagonal_allowed?(x, y) && not_backwards?(x, y) && (self.y_cord - y).abs == 1 && (self.x_cord - x).abs == 1
```
Comment (ruishine) [inline app/models/pawn.rb]: take out absolute method so we don't need not_backwards

File: app/models/pawn.rb [hunk 4]
```diff
@@ -1,5 +1,76 @@
...
+    on_board?(new_x, new_y) && passes_pawn_rules?(new_x, new_y)
+  end
+
+  def passes_pawn_rules?(x, y)
+    if two_squares_forward?(x, y)
+      return true
+    elsif one_square_forward?(x, y)
+      return true
+    elsif diagonal_move?(x, y)
+      return true
```
Comment (ruishine) [inline app/models/pawn.rb]: @ilyakrasnov are we making different methods for i.e. en passant & pawn moving diagonal to capture pieces? If so, then diagonal_move can be taken out here?

File: app/models/pawn.rb (hunk 3 above)
Comment (lindyrp) [inline app/models/pawn.rb]: The thinking with absolute values was that for the white pieces, pawns have to move in a positive way (ie, from y=1 to y=2), but black needs to move in a negative way (from y=6 to y=5). How then do we account for the difference in direction dependent on color?

File: app/models/pawn.rb (hunk 4 above)
Comment (ilyakrasnov) [inline app/models/pawn.rb]: @ruishine for now no need to consider en passant

File: app/models/pawn.rb [hunk 5]
```diff
@@ -1,5 +1,76 @@
...
+    diagonal_allowed?(x, y) && not_backwards?(x, y) && (self.y_cord - y).abs == 1 && (self.x_cord - x).abs == 1
+  end
+
+  #Checks that square is on board
+  def on_board?(x, y)
+    x >= 0 && x <= 7 && y >=0 && y <= 7
+  end
+
+  #Checks that pawn is on the originating square aka first_move_position depending on its color
+  def first_move_position?
```
Comment (ilyakrasnov) [inline app/models/pawn.rb]: What if we decide to refactor it later and black and white changes its positions for example? This piece of code will need to change.. Is there a possibility to add a flag on a piece if it was moved? Then the whole thing become: 

def moved?
    moved
end

File: db/migrate/20160708203944_add_moved_column_to_pieces.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddMovedColumnToPieces < ActiveRecord::Migration
+  def change
+    add_column :pieces, :moved, :string
```
Comment (ilyakrasnov) [inline db/migrate/20160708203944_add_moved_column_to_pieces.rb]: why not a boolean?

**Comment by Joshuavarghese527** [comment]
👍

**Pull request merged**